File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, INT_MIN);
    }
    
    int dfs(TreeNode* node, int max_val);
};

int Solution::dfs(TreeNode* node, int max_val) {
    if (!node) {
        return 0;
    }
    
    int count = (node->val >= max_val) ? 1 : 0;
    max_val = max(max_val, node->val);
    int left_count = dfs(node->left, max_val);
    int right_count = dfs(node->right, max_val);
    
    return count + left_count + right_count;
}

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, max_val):
            if not node:
                return 0
            
            count = 1 if node.val >= max_val else 0
            max_val = max(max_val, node.val)
            left_count = dfs(node.left, max_val)
            right_count = dfs(node.right, max_val)
            return count + left_count + right_count
    
        return dfs(root, float('-inf'))
